app/frontend/wallet/account-manager.ts

```diff
@@ -46,29 +49,66 @@ const AccountManager = ({
   }
 
   async function discoverAccounts() {
+    if (
+      maxAccountIndex === accounts.length - 1 || accounts.length > 0
+        ? !(await accounts[accounts.length - 1].isAccountUsed())
+        : false
+    ) {
+      return accounts
+    }
+
     const isBulkExportSupported = cryptoProvider.isFeatureSupported(
       CryptoProviderFeature.BULK_EXPORT
     )
     const shouldExplore =
       config.shouldExportPubKeyBulk && config.isShelleyCompatible && isBulkExportSupported
-    async function _discoverNextAccount(accountIndex: number) {
-      const newAccount = accounts[accountIndex] || discoverNextAccount()
-      const isAccountUsed = await newAccount.isAccountUsed()
-      if (accountIndex === accounts.length) await addNextAccount(newAccount)
-
-      return (
-        shouldExplore &&
-        isAccountUsed &&
-        accountIndex < maxAccountIndex &&
-        (await _discoverNextAccount(accountIndex + 1))
+
+    if (!shouldExplore) {
+      await exploreNextAccount()
```

Comment:
looking at the code, won't this result in discovery of additional Byron accounts when the `discoverAccounts` function is called repeatedly? If so, this doesn't seem correct to me.

Also, I find it a bit contradictory that we check for `!shouldExplorer` only to "explore" one account - looking at the code, a more appropriate name would probably be something like `shouldExploreMoreThanOne`, or define it as the opposite `shouldExploreOnlyOne` - the latter sounds a bit more natural to me. I'd rather use such verbose names than have seemingly contradictory statements like this one, leading to confusion

Reply:
Fixed.